src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java

```diff
@@ -58,146 +51,111 @@
-		this.systemEngine = loginProperties.isSystemEngine();
-		this.fireboltEngineService = fireboltEngineService;
-		this.fireboltAccountIdService = fireboltAccountIdService;
-		this.connect();
 	}
 
 	// This code duplication between constructors is done because of back reference: dependent services require reference to current instance of FireboltConnection that prevents using constructor chaining or factory method.
 	@ExcludeFromJacocoGeneratedReport
-	public FireboltConnection(@NonNull String url, Properties connectionSettings) throws SQLException {
+	protected FireboltConnection(@NonNull String url, Properties connectionSettings) throws SQLException {
 		this.loginProperties = extractFireboltProperties(url, connectionSettings);
 		OkHttpClient httpClient = getHttpClient(loginProperties);
 		ObjectMapper objectMapper = FireboltObjectMapper.getInstance();
 
-		this.fireboltAuthenticationService = new FireboltAuthenticationService(new FireboltAuthenticationClient(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients()));
-		this.fireboltGatewayUrlService = new FireboltGatewayUrlService(createFireboltAccountRetriever(httpClient, objectMapper, "engineUrl", GatewayUrlResponse.class));
-		this.httpConnectionUrl = getHttpConnectionUrl(loginProperties);
+		this.fireboltAuthenticationService = new FireboltAuthenticationService(createFireboltAuthenticationClient(httpClient, objectMapper));
+		this.httpConnectionUrl = loginProperties.getHttpConnectionUrl();
 		this.fireboltStatementService = new FireboltStatementService(new StatementClientImpl(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients()));
 
 		this.statements = new ArrayList<>();
 		this.connectionTimeout = loginProperties.getConnectionTimeoutMillis();
 		this.networkTimeout = loginProperties.getSocketTimeoutMillis();
-		this.systemEngine = loginProperties.isSystemEngine();
-		this.fireboltEngineService = new FireboltEngineService(this);
-		this.fireboltAccountIdService = new FireboltAccountIdService(createFireboltAccountRetriever(httpClient, objectMapper, "resolve", FireboltAccount.class));
+	}
+
+	protected abstract FireboltAuthenticationClient createFireboltAuthenticationClient(OkHttpClient httpClient, ObjectMapper objectMapper);
+
+	public static FireboltConnection create(@NonNull String url, Properties connectionSettings) throws SQLException {
+		return createConnectionInstance(url, connectionSettings);
+	}
 
-		this.connect();
+	private static FireboltConnection createConnectionInstance(@NonNull String url, Properties connectionSettings) throws SQLException {
+		switch(getUrlVersion(url)) {
+			case 1: return new FireboltConnectionUserPasswordAuthentication(url, connectionSettings);
+			case 2: return new FireboltConnectionServiceSecretAuthentication(url, connectionSettings);
+			default: throw new IllegalArgumentException(format("Cannot distinguish version from url %s", url));
```

Comment:
Discuss this with java compiler. It is not reachable only because getUrlVersion() in fact returns either 1 or 2. But its type is `int`, so, theoretically it can return any number.

Reply:
No, because such notation is not ready for future modifications. What are you going to di when version 3 will be introduced? Create `isNewNewVersion()`? Rename `isNewVersion()` to `isIntemediateVersion()`?